Convert to python3

azat opened the pull request:
**Please squash before merge**

Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Convert to python3. This closes #14886

Detailed description / Documentation draft:
python2 is already EOL, [since January 1, 2020](https://www.python.org/doc/sunset-python-2/)

Checklist:

- [x] scripts
- [x] stateless tests
- [x] stateful tests
- [x] integration tests
- [ ] `tests/external_models/catboost/data/build_catboost.sh` left as-is, since it uses some `ya` command (cc @KochetovNicolai)

Tech checklist:

<details>

- [x] tests (via `2to3`)
- [x] scripts (via `2to3`)
- [x] scripts (via `sed, package list)
- [x] some documentation
- [x] Dockerfile (via `sed`, package list)
- [x] pip (via `s/pip/python3 -m pip/`)
- [x] shebang's (via `sed`)
- [x] division in py3 is not c-like, use `//`

P.S I used `2to3` before for converting, and it was OK, I guess for massive converting it is to use it.
P.P.S. some commits have commands, that has been executed to update files

</details>

*This does not do any significant cleanup (but there are lots of icky code), only convert py2 to py3*

Refs: #15006
Refs: https://github.com/ClickHouse/ClickHouse/pull/10562#issuecomment-620823559

azat committed:
Run 2to3 to convert python2 to python3

$ git ls-files '*.py' | xargs 2to3 -wn

azat committed:
Update shebang to python3

azat committed:
docs: s/python2/python3

azat committed:
utils/release/release_lib.sh: require python3

azat committed:
tests: use invocation via python3

azat committed:
Convert scripts with 'env python' to python3

$ gg 'env python$' | cut -d: -f1 | sort -u | xargs 2to3 -nw

azat committed:
Update shebang

$ gg 'env python$' | cut -d: -f1 | sort -u | xargs sed -i 's/env python$/env python3/g'

azat committed:
Update Dockerfile to use python3

$ git ls-files -- */Dockerfile | xargs sed -i 's/\<python\>/python3/g'
$ git co docker/test/integration/resolver/Dockerfile

azat committed:
Require python3 packages for clickhouse-test debian package

azat committed:
Explicitly call python3

$ gg '^python ' tests/ | cut -d: -f1 | xargs sed -i 's/^python /python3 /g'

azat committed:
Update pip invocation in */Dockerfile

$ gg 'pip ' docker | cut -d: -f1 | sort -u | xargs sed -i 's/RUN pip /RUN python3 -m pip /'

azat committed:
Update some scripts to install python3 packages

azat committed:
Update tests/integration/README.md to use python3

azat committed:
clickhouse-test: use bytes instead of str to to subprocess communications

azat committed:
clickhouse-test: fix comparison of tuple() with int()

azat committed:
clickhouse-test: fix float cannot be represented as int

azat committed:
clickhouse-test: fix stdout/stderr decoding

azat committed:
Use kafka-python over kafka-python3 in yandex/clickhouse-integration-tests-runner

azat committed:
clickhouse-test: fix bytes/str decode/encode with --testname

azat committed:
Fix 00921_datetime64_compatibility with python3

azat committed:
clickhouse-test: fix diff (convert bytes to str)

azat committed:
Fix order changes after converting to python3 in 00921_datetime64_compatibility.python

- python3: {'op': ['-', '!=', '==', '< ', '<=', '> ', '>='], 'arg': ['DT', 'D', 'DT64']}
- python2: {'arg': ['DT', 'D', 'DT64'], 'op': ['-', '!=', '==', '< ', '<=', '> ', '>=']}

azat committed:
Replace rpm-confluent-schemaregistry with confluent-kafka

azat committed:
Merge remote-tracking branch 'upstream/master' into python3

Conflicts:
    tests/integration/test_mutations_with_merge_tree/test.py

azat commented:
Conflicts with #15032 (since it has few bits for converting to python3 too)

azat committed:
Merge remote-tracking branch 'upstream/master' into python3

Conflicts:
    tests/clickhouse-test

azat committed:
clickhouse-test: drop superfluous str.encode()

azat committed:
tests/*/helpers/uexpect.py: convert to python3

azat committed:
tests/queries/0_stateless/00612_http_max_query_size: update to python3

azat committed:
tests/queries/0_stateless/00565_enum_order: convert to python3

azat committed:
tests/queries/0_stateless/00411_long_accurate_number_comparison: convert to python3

azat committed:
tests: convert clickhouse_client_removed_host_parameter to python3

azat committed:
tests/queries/0_stateless/00646_url_engine: make it more reliable

azat committed:
tests/queries/0_stateless/00646_url_engine: convert to python3

azat committed:
Fix s3downloader for python3

TypeError: write() argument must be str, not bytes

azat committed:
Proper escape of \N in python3

$ gg '"\\N"' tests/integration/ | cut -d: -f1 | sort -u | xargs sed -i 's/"\\N"/"\\\\N"/g'

azat committed:
Add avro python module into intergration tests image

azat committed:
tests/queries/0_stateless/helpers/httpexpect: convert to python3

azat committed:
tests/integration/test_dictionaries_select_all/generate_dictionaries: py3

abyss7 commented:
Great effort! Looking forward to merge this PR when it's complete.

abyss7 commented:
> Conflicts with #15032 (since it has few bits for converting to python3 too)

BTW, my changes apply to pytest scripts which are unused for now in CI.

azat committed:
tests/queries/0_stateless/helpers/httpexpect: fix calling via __main__

azat committed:
tests/queries/0_stateless/helpers/httpclient: set timeout to avoid hang

azat committed:
Fix 00411_long_accurate_number_comparison under python3

azat commented:
>Looking forward to merge this PR when it's complete.

@abyss7 BTW I'm open to any contributions here, since this should be merged ASAP, otherwise there will be more and more conflicts

abyss7 commented on tests/clickhouse-test:
```diff
@@ -130,7 +130,7 @@ def run_single_test(args, ext, server_logs_level, client_options, case_file, std
         database = 'test_{suffix}'.format(suffix=random_str())
 
         clickhouse_proc_create = Popen(shlex.split(args.client), stdin=PIPE, stdout=PIPE, stderr=PIPE)
```
Maybe `Text=True` argument will help to avoid encoding of input and output lines

azat replied:
Indeed

nikitamikhaylov commented:
> > Looking forward to merge this PR when it's complete.
> 
> @abyss7 BTW I'm open to any contributions here, since this should be merged ASAP, otherwise there will be more and more conflicts

@azat  I will probably help with some of integration tests.

nikitamikhaylov committed:
copier & client

(cherry picked from commit f8a03f41bad841a41ca1c1018ebb1516de896567)

azat commented:
Conflicts with #15110

azat commented:
Integration tests for 9f891c42ec7fd6a31668d125cc874493a3c72e22 commit definitely hanged, can someone obtain logs from them?

azat committed:
clickhouse-test: use Popen(text=True) over manual encode()/decode()

azat committed:
clickhouse-test: fix possible attempt of decoding str instead of bytes

azat committed:
tests/helpers/uexpect: replace errors

Otherwise the following can happen (due to progress?) [1]:

2020-09-21 03:11:45 00979_live_view_watch_live_with_subquery:                               [ FAIL ] 121.36 sec. - return code 1
2020-09-21 03:11:45 Exception in thread Thread-1:
2020-09-21 03:11:45 Traceback (most recent call last):
2020-09-21 03:11:45   File "/usr/lib/python3.7/threading.py", line 926, in _bootstrap_inner
2020-09-21 03:11:45     self.run()
2020-09-21 03:11:45   File "/usr/lib/python3.7/threading.py", line 870, in run
2020-09-21 03:11:45     self._target(*self._args, **self._kwargs)
2020-09-21 03:11:45   File "/usr/share/clickhouse-test/queries/0_stateless/helpers/uexpect.py", line 201, in reader
2020-09-21 03:11:45     data = os.read(out, 65536).decode()
2020-09-21 03:11:45 UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe2 in position 107: unexpected end of data
2020-09-21 03:11:45
2020-09-21 03:11:45 Traceback (most recent call last):
2020-09-21 03:11:45   File "/usr/share/clickhouse-test/queries/0_stateless/helpers/uexpect.py", line 170, in read
2020-09-21 03:11:45     data += self.queue.get(timeout=timeleft)
2020-09-21 03:11:45   File "/usr/lib/python3.7/queue.py", line 178, in get
2020-09-21 03:11:45     raise Empty
2020-09-21 03:11:45 _queue.Empty
2020-09-21 03:11:45
2020-09-21 03:11:45 During handling of the above exception, another exception occurred:
2020-09-21 03:11:45
2020-09-21 03:11:45 Traceback (most recent call last):
2020-09-21 03:11:45   File "/usr/share/clickhouse-test/queries/0_stateless/helpers/uexpect.py", line 144, in expect
2020-09-21 03:11:45     data = self.read(timeout=timeleft, raise_exception=True)
2020-09-21 03:11:45   File "/usr/share/clickhouse-test/queries/0_stateless/helpers/uexpect.py", line 178, in read
2020-09-21 03:11:45     raise TimeoutError(timeout)
2020-09-21 03:11:45 uexpect.TimeoutError: Timeout 120.000s
2020-09-21 03:11:45
2020-09-21 03:11:45 During handling of the above exception, another exception occurred:
2020-09-21 03:11:45
2020-09-21 03:11:45 Traceback (most recent call last):
2020-09-21 03:11:45   File "/usr/share/clickhouse-test/queries/0_stateless/00979_live_view_watch_live_with_subquery.py", line 43, in <module>
2020-09-21 03:11:45     client1.expect(r'%d.*%d' % (21+i, 3+i) + end_of_block)
2020-09-21 03:11:45   File "/usr/share/clickhouse-test/queries/0_stateless/helpers/uexpect.py", line 151, in expect
2020-09-21 03:11:45     raise exception
2020-09-21 03:11:45 uexpect.ExpectTimeoutError: Timeout 120.000s for '45.*27.*\\r\\n.*\\r\\n' buffer '\r\x1b[1;34m←\x1b[0m Progress: 20.00 rows, 320.00 B (45.48 rows/s., 727.70 B/s.) \x1b[K\r\x1b[1;35m↖\x1b[0m Progress: 26.00 rows, 416.00 B (59.12 rows/s., 945.94 B/s.) \x1b[K'
2020-09-21 03:11:45
2020-09-21 03:11:45 , result:
2020-09-21 03:11:45

  [1]: https://clickhouse-test-reports.s3.yandex.net/15007/9f891c42ec7fd6a31668d125cc874493a3c72e22/functional_stateless_tests_(address)/test_run.txt.out.log

azat committed:
tests/integration: make http_query() python3 compatible

azat committed:
test_storage_kafka: fixes for python3

- add value_serializer for KafkaProducer
- add missing MessageSerializer into test_storage_kafka
- avro_confluent_message() return bytes not str
- fix binary data
- fix kafka_produce_protobuf_messages

azat committed:
tests/integration: use text output for query execution via client

azat committed:
Use universal_newlines=True over text=True

Since in docker images there is python 3.6, while text=True since 3.7

azat committed:
tests/integration: difflib.unified_diff() does not accept None

alexey-milovidov commented:
Stopeed after 4.5 hours, 
https://drive.google.com/file/d/1PNrqvhYkJV8pbgnVDUuM3VeGIAcRNJJY/view?usp=sharing

azat committed:
integration: fix hdfs_api usage

azat committed:
Fix test_http_handlers_config for python3 (bytes not str)

azat committed:
Fix test_zookeeper_config for python3

azat commented:
>Integration tests (asan) — fail: 88, passed: 676, error: 77
>Integration tests (release) — fail: 86, passed: 676, error: 79

At least now it does not timedout

azat committed:
tests/integration: simplity CommandRequest by chaning mode for file (py3)

azat committed:
integration/hdfs_api: fix check_output() by adding universal_newlines=True (for py3)

azat committed:
integration/helpers/cluster: fix http_query() for py3

azat committed:
Fix test_ttl_move for python3

azat committed:
integration/helpers/client: open file in binary and the decode for py3

    def decode(self, input, final=False):
>       return codecs.ascii_decode(input, self.errors)[0]
E       UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 7: ordinal not in range(128)

azat committed:
integration/helpers/cluster: fix exec_in_container() py3

azat committed:
test_mysql_protocol py3 (bytes to str)

azat committed:
test_postgresql_protocol py3 (bytes to str)

azat committed:
integration/helpers/http_server: fix for py3 (str and bytes)

azat committed:
Fix test_storage_kafka

azat committed:
Fix test_no_local_metadata_node

azat committed:
Fix test_replicated_parse_zk_metadata

azat committed:
Fix test_storage_kafka/test.py::test_kafka_virtual_columns2

azat committed:
Fix test_storage_kafka/test.py::test_kafka_virtual_columns2

azat committed:
Bump ci

azat commented:
> Integration tests (asan) — No tests found for some reason!

By some reason it uses version of some scripts from upstream/master (I guess not all images are versioned?)

```
conftest.py:1: in <module>
    from helpers.test_tools import TSV
/usr/lib/python3.6/ast.py:35: in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
E     File "/ClickHouse/tests/integration/helpers/test_tools.py", line 76
E       print "contains_in_log_with_retry retry {} exception {}".format(i + 1, ex)
E                                                              ^
E   SyntaxError: invalid syntax
```

Can this be fixed/avoided?

azat committed:
Merge remote-tracking branch 'upstream/master' into python3

Conflicts:
        tests/clickhouse-test
        tests/integration/helpers/cluster.py
        tests/integration/test_dictionaries_all_layouts_and_sources/test.py
        tests/integration/test_replicated_merge_tree_s3/test.py

azat committed:
Fix tests/integration/helpers/test_tools

azat committed:
Fix test_dictionaries_all_layouts_separate_sources

azat committed:
test_system_queries: remove debug output

azat committed:
Fix test_allowed_client_hosts

azat committed:
Fix test_config_substitutions

azat committed:
Fix test_allowed_client_hosts

azat committed:
Fix test_host_ip_change

alesapin commented:
Check `Integration tests flaky check (asan)` can be ignored. We have a lot of flaky integration tests and we will fix them gradually.

azat committed:
Fix test_storage_s3

azat committed:
Fix test_multiple_disks

azat committed:
Fix test_backup_restore

azat committed:
Fix test_storage_hdfs

azat committed:
Fix test_partition

azat committed:
Fix test_distributed_format

azat committed:
Fix test_reloading_storage_configuration

azat committed:
integration: exec_in_container() return str

azat committed:
Fix hdfs_api

azat committed:
Fix test_storage_s3

azat committed:
Replace slow get_random_string() with randomPrintableASCII() in test_ttl_move

After converting to python3 get_random_string() became even slower due
to decode() and breaks the test

azat committed:
Fix test_allowed_client_hosts

azat committed:
Fix exec_in_container()

azat committed:
Fix get_container_logs()

azat committed:
Fix hdfs_api.py::read_gzip_data

azat committed:
Fix hdfs_api.py::write_gzip_data()

azat committed:
Fix test_distributed_ddl

azat committed:
Disable logging for live view tests

azat committed:
Fix test_send_crash_reports

azat committed:
Merge remote-tracking branch 'upstream/master' into python3

Conflicts:
    tests/integration/test_mysql_protocol/test.py

azat committed:
Improve error message in test_send_crash_reports

azat committed:
Fix test_ttl_move

azat committed:
Update test_send_crash_reports

azat committed:
Fix hdfs_api

azat committed:
Fix test_replicated_merge_tree_s3

azat committed:
Fix test_dictionary_custom_settings/http_server

azat committed:
Fix test_distributed_ddl

azat committed:
Fix http_query() auth

azat commented:
If upstream/master will not go ahead then this maybe enough

qoega commented on docs/ru/development/build.md:
```diff
@@ -1 +0,0 @@
-../../en/development/build.md
```
Why did you convert it to copy?

azat replied:
Nice catch, by mistake

azat replied:
Now there is 0 such bits:

```
$ git di upstream/master... | grep -c 'deleted file'
0
```

qoega commented on tests/queries/0_stateless/00921_datetime64_compatibility.reference:
```diff
@@ -1293,98 +1292,99 @@ Code: 43: Illegal types of arguments (UInt64, Date) of function greater.
+SELECT toUInt64(1) >= N
 
-Code: 43: Illegal types of arguments (Int64, Date) of function greater.
+Code: 43: Illegal types of arguments (UInt64, Date) of function greaterOrEquals.
```
It is not trivial to check diff in this reference. I hope you've tried to check that nothing was broken or lost.

azat replied:
> It is not trivial to check diff in this reference.

Indeed (test is tricky)

> I hope you've tried to check that nothing was broken or lost.

But yes, the problem basically is in the order, and it has been fixed separately in 5228c5b8aa3ea698e4050e960c84835dacb00e62:

    Fix order changes after converting to python3 in 00921_datetime64_compatibility.python
    
    - python3: {'op': ['-', '!=', '==', '< ', '<=', '> ', '>='], 'arg': ['DT', 'D', 'DT64']}
    - python2: {'arg': ['DT', 'D', 'DT64'], 'op': ['-', '!=', '==', '< ', '<=', '> ', '>=']}

qoega commented on tests/integration/test_send_crash_reports/fake_sentry_server.py:
```diff
@@ -1,25 +1,26 @@
-            elif self.headers.get("content-length") < 3000:
-                f.write("INCORRECT_CONTENT_LENGTH")
-            elif '"http://6f33034cfe684dd7a3ab9875e57b1c8d@localhost:9500/5226277"' not in post_data:
+            elif int(content_length) < 200:
```
Why is 3000 changed to 200?

azat replied:
Because the payload is not that huge, and the test was always broken, since it compares `str` with number:

```python
$ python2 -c "print('263' > 3000)"
True
```

And it is fine in python2 but not in python3:

```python
$ python3 -c "print('263' > 3000)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: '>' not supported between instances of 'str' and 'int'
```

excitoon commented on docs/tools/test.py:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 # -*- coding: utf-8 -*-
```
I'd remove these lines as well as it's the default encoding.

azat replied:
Why do you think that shebang should be removed?
I'm ok with removing the coding

azat committed:
Fix docs/ru/development/build.md

azat committed:
docs/tools/test.py: drop coding

azat committed:
Merge remote-tracking branch 'upstream/master' into python3

New tests:

$ git di ...upstream/master --stat tests/integration/
 tests/integration/helpers/cluster.py                                        |  40 +++++++++++++++++++++++++++++++++++++---
 tests/integration/runner                                                    |   4 +++-
 tests/integration/test_compression_nested_columns/__init__.py               |   0
 tests/integration/test_compression_nested_columns/test.py                   |  68 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 tests/integration/test_mysql_database_engine/test.py                        |  33 +++++++++++++++++++++++++++++++++
 tests/integration/test_storage_kerberized_kafka/__init__.py                 |   0
 tests/integration/test_storage_kerberized_kafka/clickhouse_path/EMPTY_DIR   |   0
 tests/integration/test_storage_kerberized_kafka/configs/kafka.xml           |  26 ++++++++++++++++++++++++++
 tests/integration/test_storage_kerberized_kafka/configs/log_conf.xml        |  11 +++++++++++
 tests/integration/test_storage_kerberized_kafka/kerberos_image_config.sh    | 132 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 tests/integration/test_storage_kerberized_kafka/secrets/broker_jaas.conf    |  14 ++++++++++++++
 tests/integration/test_storage_kerberized_kafka/secrets/krb.conf            |  22 ++++++++++++++++++++++
 tests/integration/test_storage_kerberized_kafka/secrets/zookeeper_jaas.conf |  14 ++++++++++++++
 tests/integration/test_storage_kerberized_kafka/test.py                     | 146 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 14 files changed, 506 insertions(+), 4 deletions(-)

azat committed:
Fix test_storage_kerberized_kafka

azat committed:
Rename serializer in test_storage_kafka

azat committed:
Fix test_storage_kerberized_kafka

azat committed:
Merge remote-tracking branch 'upstream/master' into python3

$ git di ...upstream/master --stat tests/integration/
 tests/integration/test_merge_tree_s3/test.py  |   3 ---
 tests/integration/test_rename_column/test.py  |   1 -
 tests/integration/test_storage_s3/test.py     | 121 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 tests/integration/test_ttl_replicated/test.py |   1 -
 4 files changed, 121 insertions(+), 5 deletions(-)

azat committed:
Fix test_storage_s3

azat committed:
Merge remote-tracking branch 'upstream/master' into python3

azat committed:
Make live view tests more stable by increasing buffer size

azat committed:
Merge remote-tracking branch 'upstream/master' into python3

* upstream/master:
  Throw an error when a single parameter is passed to ReplicatedMergeTree instead of ignoring it

azat commented:
>Integration tests (release) — fail: 1, passed: 901, error: 0

`test_distributed_over_live_view/test.py::TestLiveViewOverDistributedSuite::test_distributed_over_live_view_order_by_node[lv_over_distributed_table-node0]` is flacky (fixed by increasing buffer size for now)
Apart from this, other tests are ok (except for `Integration tests flaky check`, but they are not related to this conversion)
